File: fsharp/allergies/Allergies.fs
module Allergies

open System

type Allergen =
    | Eggs
    | Peanuts
    | Shellfish
    | Strawberries
    | Tomatoes
    | Chocolate
    | Pollen
    | Cats
    
let allergens = [Allergen.Eggs; Allergen.Peanuts; Allergen.Shellfish; Allergen.Strawberries;
                Allergen.Tomatoes; Allergen.Chocolate; Allergen.Pollen; Allergen.Cats]

let powerOf2 n = 1 <<< n

let getIndex allergen = allergens |> List.findIndex (fun x -> x = allergen)

let allergicTo codedAllergies allergen =
    let inx = getIndex allergen
    let code = powerOf2 inx
    (codedAllergies &&& code) > 0

let listBuilder currList codedAllergies allergen =
    if allergicTo codedAllergies allergen then allergen :: currList
    else currList
    
let list codedAllergies =
    ([], [0..7])
    ||> Seq.fold (fun currList inx -> (listBuilder currList codedAllergies allergens.[inx]))
    |> List.rev
            

File: r/allergies/allergies.R
allergens <- c( eggs = 1,
                peanuts = 2,
                shellfish = 4,
                strawberries = 8,
                tomatoes = 16,
                chocolate = 32,
                pollen = 64,
                cats = 128 )

allergy <- function(num) {
  num <- num %% 256
  names(allergens[bitwAnd(allergens, num) != 0])
}

allergic_to <- function(allergy_object, allergy) allergy %in% allergy_object

list_allergies <- function(allergy_object) allergy_object
